contrib/jmx-metrics/src/main/groovy/io/opentelemetry/contrib/jmxmetrics/InstrumentHelper.groovy

```diff
@@ -89,21 +96,48 @@ class InstrumentHelper {
             return
         }
 
+        // Observer instruments need to have a single callback set, so pool all update
+        // operations in a list of closures per instrument to be executed after all values
+        // are established, potentially as the callback.
+        def instToUpdates = [:]
+
         [mbeans, values].transpose().each { mbean, value ->
             if (value instanceof CompositeData) {
                 value.getCompositeType().keySet().each { key ->
                     def val = value.get(key)
                     def updatedInstrumentName = "${instrumentName}.${key}"
                     def labels = getLabels(mbean, labelFuncs)
                     def inst = instrument(updatedInstrumentName, description, unit)
+                    println "InstrumentHelper.update (composite) - ${inst}"
                     logger.fine("Recording ${updatedInstrumentName} - ${inst} w/ ${val} - ${labels}")
-                    updateInstrumentWithValue(inst, val, labels)
+                    if (!instToUpdates.containsKey(inst)) {
```

Comment:
Nit: looks like you could factor out these four lines with line 123-126 if you wanted.

Reply:
Since this section is in an inner iterator I think I'd have to store everything in an additional structure for final `instToUpdates` modification at the end and I'm not sure that's worth the reduced redundancy.  I'm generally a "rule of three" follower though...